---
title: Animated plots and single-cell RNA-seq analysis
author: Jacques Serizay
date: 2020-09-03
slug: gganimate
comments: false
showpagemeta: true
showcomments: false
categories:
    - R
    - Bioconductor
    - 'Data Viz'
    - 'Single cell'
tags:
    - R
    - Bioconductor
    - 'Data Viz'
    - 'Single cell'
output:
    blogdown::html_page:
        highlight: "tango" 
        toc: true
        toc_depth: 1
        number_sections: false
        fig_width: 12
---

I know I am probably being over-enthousiastic about it, but I keep learning about `ggplot2` features and I can't get enough of it. Lately, I have discovered the possibilities of animating ggplots based on aesthetics of interest, and as usual, this is as simple as a single line of code... 

In this post, I will rely on some single-cell RNA-seq data (from a 10X Genomics sequencing run) containing gene expression for ~500 mouse cycling Radial Glial Progenitors (RGPs). 

## Loading the single-cell dataset as a SingleCellExperiment object

```{r}
library(SingleCellExperiment)
cyclingCells <- readRDS('data/cyclingCells.rds')
cyclingCells
head(rowData(cyclingCells))
head(colData(cyclingCells))
```

## Embedding cells in lower dimensional space

```{r}
library(scran)
library(scater)
set.seed(100)
cycling_genes <- c(
    "Aurkb", "Ccnb1", "Ccnb2", "Ccnd1", "Ccnd2", "Cdc20", "Cdca3",
    "Cdk1", "Cdkn3", "Cenpa", "Cenpe", "Cenpf", "Cenpq", "Cks2",
    "Gmnn", "H2afx", "H2afz", "Hist1h1b", "Hist1h1c", "Hist1h1e",
    "Hist1h2ae", "Hist1h2bc", "Mcm3", "Mcm4", "Mcm6", "Mki67", "Prc1",
    "Serpine2", "Top2a", "Ube2c", "Ube2s"
)
deconv_var <- modelGeneVarByPoisson(cyclingCells)
cyclingCells <- denoisePCA(
    cyclingCells, 
    subset.row = cycling_genes, 
    technical = deconv_var, 
    min.rank = 50
)
reducedDim(cyclingCells, "force") <- igraph::layout_with_fr(
    buildSNNGraph(
        cyclingCells, 
        use.dimred = "PCA", 
        subset.row = cycling_genes
    )
)
```

We can plot the cells embedded in a lower dimensional space with a function from my package of single-cell utilities: 

```{r}
library(SCTools)
# ----- Plotting embedding in a lower dim. space ----- #
SCTools::plotEmbedding(
    cyclingCells, 
    "detected", 
    "force"
)
```

We can also color by levels of gene expression: 

```{r}
SCTools::plotEmbedding(
    cyclingCells, 
    "Ccne2", 
    "force"
)
SCTools::plotEmbedding(
    cyclingCells, 
    "Ccnb2", 
    "force"
)
```

Finally, we can plot several genes at once: 

```{r}
genes <- c('Ccnd1', 'Ccne2', 'Ccne1', 'Ccna2', 'Ccnb1', 'Ccnb2', 'Ccna1')
SCTools::plotEmbedding(
    cyclingCells, 
    c('Ccnd1', 'Ccne2', 'Ccne1', 'Ccna2', 'Ccnb1', 'Ccnb2', 'Ccna1'), 
    "force", 
    theme.args = theme(legend.position = 'none')
)
```

## Putting all these plots in 1 GIF

```{r}
library(tidyverse)
library(magrittr)
df <- data.frame(cell = cyclingCells$Barcode)
for (gene in genes) {
    expr <- logcounts(cyclingCells)[gene, ] %>% bindByQuantiles(q_low = 0.05, q_high = 0.95)
    df[, paste0(gene, "_expr")] <- expr
}
df %<>% 
    mutate(
        X = reducedDim(cyclingCells, 'force')[, 1],
        Y = reducedDim(cyclingCells, 'force')[, 2]
    ) %>% 
    gather("gene", "expr", -cell, -X, -Y) %>% 
    mutate(gene = factor(gene, levels = paste0(genes, "_expr"))) %>%
    select(X, Y, gene, expr)
head(df)
p <- ggplot(df, aes_string(x = "X", y = "Y", fill = "expr")) + 
    geom_point(pch = 21, alpha = 0.5, col = '#bcbcbc', stroke = 0.2) + 
    theme_void() + theme(legend.position = 'none') +
    scale_fill_distiller(palette = 'YlOrBr', direction = 1) + 
    coord_fixed((max(df[, "X"])-min(df[, "X"]))/(max(df[, "Y"])-min(df[, "Y"]))) + 
    labs(y = "force 2", x = "force 1", fill = '') + 
    theme(
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), 
        axis.ticks = element_blank()
    ) + 
    gganimate::transition_states(
        gene,
        transition_length = 1,
        state_length = 1
    )
```

All of this is actually wrapped into a single function in `SCTools`: 

```{r}
genes <- c('Ccnd1', 'Ccne2', 'Ccne1', 'Ccna2', 'Ccnb1', 'Ccnb2', 'Ccna1')
colData(cyclingCells)$force_X <- reducedDim(cyclingCells, 'force')[, 1]
colData(cyclingCells)$force_Y <- reducedDim(cyclingCells, 'force')[, 2]
gif <- plotAnimatedEmbedding(
    cyclingCells, 
    dim = "force", 
    genes = genes, 
    theme.args = theme_void() + theme(legend.position = 'none')
) %>% gganimate::animate(
    duration = 2, fps = 20, 
    width = 400, height = 400, res = 150, 
    renderer = gganimate::gifski_renderer()
)
gif
```
