---
title: Why is Bioconductor awesome
author: Jacques Serizay
date: 2020-04-08
slug: awesome_bioconductor
comments: false
showpagemeta: true
showcomments: false
categories:
    - R
    - Bioconductor
    - Resources
tags:
    - R
    - Bioconductor
    - Resources
output:
    blogdown::html_page:
        highlight: "tango" 
        toc: true
        toc_depth: 1
        number_sections: false
        fig_width: 12
---

## I love Bioconductor 

Since its early ages, Bioconductor has gone very far. I've been on board for the past 7-ish years (I think), I use it every day, and every day, I think about how much I love it. 

## Some reasons... 

I love Bioconductor for many reasons, some of them being `GenomicRanges`, `SummarizedExperiment`, `DESeq2`, and of course [the amazing team](https://www.bioconductor.org/about/core-team/). 

But today, one reason which I want to discuss in more depth is the `Annotation Hub`. This hub provides access to tons of resources, particularly if you work in human genomics. One reason why it's so useful in human is because major consortia like Encode and Roadmap Epigenomics Project put their data there. So you end up with tons of stuff available, if you know what you are looking for!

## An example: finding genes undergoing chromatin state transition during cell differentiation

The Roadmap Epigenomics Project (REP) has characterized chromatin states in tens of epigenomes (that's literally the title of their paper: [Integrative analysis of 111 reference human epigenomes](https://www.nature.com/articles/nature14248)!).  

And guess what, all the results are available in the `AnnotationHub`! Here is a glimpse at some of the data hosted there: 

```{r}
library(tidyverse)
library(AnnotationHub)
ah <- AnnotationHub()
ah_human_epigenome <- query(ah, c("Homo sapiens", "EpigenomeRoadMap"))
table(ah_human_epigenome$rdataclass)
head(query(ah_human_epigenome, 'BigWigFile')$title, 50)
```

Wow, almost 10K of bigwig files. Neat!

What about metadata? It is available in the accession `AH41830`.

```{r}
epigenome_metadata <- ah_human_epigenome[['AH41830']]
epigenome_metadata[1:10, 1:10]
```

Cool. Next, we are going to focus on chromatin state segmentation. What's available there?

```{r}
query(ah_human_epigenome, c("chromhmmSegmentations", "cell line"))
```

Looks like 5 different epigenomes were segmented into chromatin states. The ones that we are interested in are those from H1 embryonic stem cells (H1) or from Neural Progenitor Cells (NPCs) derived from these H1 cells. 

```{r}
H1_states <- query(ah_human_epigenome, c("chromhmmSegmentations", "ESC.H1"))[[1]]
NPC_states <- query(ah_human_epigenome, c("chromhmmSegmentations", "ESDR.H1.NEUR.PROG"))[[1]]
H1_states <- resize(H1_states, 1, fix = 'center') # Resizing to make it easier to find overlaps
NPC_states <- resize(NPC_states, 1, fix = 'center') # Resizing to make it easier to find overlaps
```

Let's find out how loci are changing between H1 and NPC cultures:

```{r}
states <- c(
    "Active TSS", 
    "Flanking Active TSS", 
    "Bivalent/Poised TSS", 
    "Enhancers",
    "Flanking Bivalent TSS/Enh", 
    "Strong transcription",
    "Transcr. at gene 5' and 3'", 
    "Weak transcription",
    "Bivalent Enhancer", 
    "Genic enhancers",
    "Quiescent/Low", 
    "Heterochromatin", 
    "Weak Repressed PolyComb", 
    "Repressed PolyComb", 
    "ZNF genes & repeats"
)
df <- data.frame(
    stateInH1 = factor(H1_states$name, levels = states),
    stateInNPCs = factor(NPC_states[nearest(H1_states, NPC_states)]$name, levels = states)
)
df %>% 
    group_by(stateInH1, stateInNPCs) %>% 
    summarize(count = n()) %>% 
    mutate(label = ifelse(count > 1000, count, '')) %>%
    ggplot(aes(x = stateInH1, y = stateInNPCs, fill = count, label = label)) + 
    geom_tile(col = "black") + 
    geom_text(size = 1.5) + 
    theme_bw() + 
    scale_fill_gradientn(colours = c('white', 'orange', 'darkred')) + 
    theme(
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), 
        panel.border = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        axis.ticks = element_blank()
    ) + 
    labs(
        x = "Chromatin states in H1 cells", 
        y = "Chromatin states in H1-derived\nneural progenitor cells", 
        fill = '#'
    ) + 
    coord_fixed()
```

Using the `tidyverse` to plot the transitions, it's clear that while most of the segmented genomic loci keep their functional annotation from ESCs to NPCs, there are other loci which change from inactive to activate chromatin states, and vice versa.  

For instance, let's focus on the loci undergoing changes from a `Repressed PolyComb` state to a `Enhancers` state:

```{r}
transitioning_loci <- H1_states[df$stateInH1 == 'Repressed PolyComb' & df$stateInNPCs == 'Enhancers']
transitioning_loci
```

Now, let's fetch the genes associated to these loci. This is easily done, again, by retrieving informatin from `AnnotationHub`!  

I will retrieve the human gene model from `AnnotationHub`: 

```{r}
human_gtf <- query(ah, c("Homo sapiens", "release-75"))[[1]]
seqlevelsStyle(human_gtf) <- "UCSC"
human_gtf <- keepStandardChromosomes(human_gtf, pruning.mode = 'coarse')
human_genes <- human_gtf[human_gtf$type == 'gene' & human_gtf$gene_biotype == 'protein_coding']
human_TSSs <- resize(human_genes, 1, 'start')
```

I can then retrieve the closest gene for each of the transitioning loci, and filter the genes that are close to these transitioning loci. 

```{r}
transitioning_loci$nearest_gene <- human_TSSs[nearest(transitioning_loci, human_TSSs)]$gene_id
transitioning_loci$distance_to_nearest_gene <- mcols(distanceToNearest(transitioning_loci, human_TSSs))$distance
genes <- unique(transitioning_loci$nearest_gene[transitioning_loci$distance_to_nearest_gene <= 1000]) 
```

Finally, a quick GO over-representation analysis will reveal whether these few genes are indeed involved in any neural process. Once again, I can retrieve ontology from human `OrgDb` using `AnnotationHub`: 

```{r}
org <- query(ah, c('homo sapiens', 'orgdb'))[[1]]
res <- clusterProfiler::enrichGO(
    genes, 
    OrgDb = org,
    ont = 'CC',
    keyType = 'ENSEMBL', 
    pvalueCutoff = 0.05, 
    qvalueCutoff = 0.05
)
res@result %>% dplyr::select(Description, pvalue, p.adjust, Count) %>% arrange(p.adjust) %>% filter(p.adjust < 0.05)
```

Clearly, some of the genes undergoing chromatin state transition are involved in neurone cell structures!

